TRC description: mandate normalization, introduce language tag and localizedDescriptions (#137)
scionassociation/scion-cppki_I-D: draft-dekater-scion-pki.md

### `description` {#description}

The `description` field contains a UTF-8 encoded string that describes the ISD. The text MUST be formatted in accordance with "Net-Unicode" {{RFC5198}} to ensure consistent normalization. It SHOULD NOT be empty.

> The description MUST be in English. If multiple languages are provided, each language section SHOULD be identified by a language tags as defined in {{BCP47}}.

Reviewer: I don't think this can be a MUST for English. We have examples of TRCs that are not in English already.
E.g. ISD70, ISD72, ISD 76. It also feels a bit like an overreach. We want SCION to be adopted across the world. 

There is a logical inconsistency, fwiw. If the description is English, it cannot have multiple languages.

Author: > The description MUST be in English  

This comes from -00 and likely comes from some old docs. I see your point, let's drop it

### `description` {#description}

> The `description` field contains a UTF-8 encoded string that describes the ISD. The text MUST be formatted in accordance with "Net-Unicode" {{RFC5198}} to ensure consistent normalization. If present, it MUST NOT be empty.

Reviewer 2: This is already implied by the ASN.1 description, which specifies a minimum size of 1. Do we need an additional MUST here?
Id suggest removing "If present, it MUST NOT be empty."

Author: A field containing just a whitespace would be parsed and OK according to the ASN.1 module I don't think that is the point, so I recommend leave this wording where it is.

Author: Or at least leaving a SHOULD: 
```suggestion
The `description` field contains a UTF-8 encoded string that describes the ISD. The text MUST be formatted in accordance with "Net-Unicode" {{RFC5198}} to ensure consistent normalization. If present, it SHOULD NOT be empty.
```

Reviewer 2: > A field containing just a whitespace would be parsed and OK according to the ASN.1 module I don't think that is the point, so I recommend leave this wording where it is.

I think what you really want to specify here is that the content should be **sensible**, but I am not sure that can specified with normative language.

It is also debatable whether a string with a single character is empty, regardless of whether this is a whitespace or not. If you want to prevent a whitespace, a more precise norm would be: `If present, it MUST NOT contain only whitespaces.` ...?

That would still allow descriptions like `123456`, so I think this cannot be resolved with normative language and I suggest to remove it.

Author: I would still encourage deployers to put something in that field.. I don't think we can be the protocol police and avoid that people put 12345 in it, but I think we should at least give some human-language guidance, and I see no harm in having a SHOULD not be empty. It becomes a lot clearer than just reading the 1 in the ASN.1 module

Reviewer 2: Okay :+1:

> When this field contains languages other than English, the corresponding language tag SHOULD explicitly identified in the `descriptionLanguage` field (see ()[#langtag]).

Reviewer 2: ```suggestion
When this field contains languages other than English, the corresponding language SHOULD be identified explicitly in the `descriptionLanguage` field (see ()[#langtag]).
```

> Multi-language TRCs SHOULD use the `localizedDescriptions` field instead. Either the `description` or the `localizedDescriptions`field MUST be present.

Reviewer 2: ```suggestion
Multi-language TRCs SHOULD use the `localizedDescriptions` field instead of the `description` field. Either the `description` or the `localizedDescriptions`field MUST be present.
```

It is a bit unclear what "instead" refers to. Does it mean that one should use ether "description" or "localizedDescriptions", but not both?

A certificate's type (voting or root) is specified in the `extKeyUsage` extension of the certificate, by means of the SCION-specific attributes `id-kp-regular`, `id-kp-sensitive`, and `id-kp-root`, respectively. For more information, see [](#ext-key-usage-ext).

### `localizedDescriptions` {#description-multilang}

The `localizedDescriptions` field provides an optional mechanism for including multilingual descriptions.
It consists of a sequence of `LocalizedText` structures, each containing:

- `languageTag`: specifies the description's language. It MUST use language tags according to {{BCP47}}.

> - `text`: contains the localized description. It MUST be formatted in accordance with "Net-Unicode" {{RFC5198}}.

Reviewer 2: ```suggestion
- `language`: specifies the description's language. It MUST use language tags according to {{BCP47}}.
- `content`: contains the localized description. It MUST be formatted in accordance with "Net-Unicode" {{RFC5198}}.
```

### `description` {#description}

> The `description` field contains a UTF-8 encoded string that describes the ISD. The text SHOULD be formatted in accordance with "Net-Unicode" {{RFC5198}} to ensure consistent normalization. It SHOULD NOT be empty.

Reviewer: We can make it a MUST if you want. All existing TRCs are in that form

Author: ```suggestion
The `description` field contains a UTF-8 encoded string that describes the ISD. The text MUST be formatted in accordance with "Net-Unicode" {{RFC5198}} to ensure consistent normalization. It SHOULD NOT be empty.
```

### `description` {#description}

The `description` field contains a UTF-8 encoded string that describes the ISD. The text MUST be formatted in accordance with "Net-Unicode" {{RFC5198}} to ensure consistent normalization. It SHOULD NOT be empty.
The description MUST be in English. If multiple languages are provided, each language section SHOULD be identified by a language tags as defined in {{BCP47}}.

> When using language tags, each tag SHOULD be enclosed in square brackets (e.g., [en], [de-CH]) at the start of its respective section.

Author: @oncilla how about this?

Reviewer: I don't think this works well, and feels like an after thought. It also makes all our TRCs go against the SHOULD recommendation.

How about the following:
We add a backward/forward compatible additional field:
`descriptions` (optional, at the end for DER, with explicit tag) that is a list of description and language.
Multi-language TRCs should use the descriptions field. No further restrictions on the description field itself.
(We could even go further and say all TRCs should use the descriptions field rather than description)

I need to double check that this is actually backward and forward compatible.

For inspiration: `LogotypeImageInfo` https://datatracker.ietf.org/doc/html/rfc9399#name-loyalty-logotype

Additionally, we can add a descriptionLanguageTag similar to: `taTitleLangTag` https://datatracker.ietf.org/doc/html/rfc5914#section-2.4

Author: I updated the text and ASN.1 module to introduce two new optional texts. What do you think?

    notAfter           GeneralizedTime
}

LocalizedText ::= SEQUENCE {
    language        PrintableString,

>     content         UTF8String (SIZE (0..8192))

Reviewer 2: See discussion on trc.asn1 above.

Author: Le't work on the .asn1 file firt, then I copy paste it into the draft at the end

}
TRCPayload ::= SEQUENCE {
    version               TRCFormatVersion,
    iD                    TRCID,
    validity              TRCValidity,
    gracePeriod           INTEGER,
    noTrustReset          BOOLEAN,
    votes                 SEQUENCE SIZE (0..2047) OF INTEGER (0..4095),
    votingQuorum          INTEGER (1..2047),
    coreASes              SEQUENCE OF ASN,
    authoritativeASes     SEQUENCE OF ASN,
    description           UTF8String (SIZE (1..8192)) OPTIONAL,
    certificates          SEQUENCE SIZE (1..4095) OF Certificate,

>     localizedDescriptions [0] SEQUENCE SIZE (1..MAX) OF LocalizedText OPTIONAL,

Reviewer 2: See discussion on trc.asn1 above.
